### PR TITLE
tests: sockets: add test for socket api tls extensions

### LINKED ISSUE #30936: tests: sockets: tcp: add a tls test
**Is your enhancement proposal related to a problem? Please describe.**
Currently I'm attempting to use mbedTLS in a custom application and am creating a test to verify TLS in CI. The test is multi-threaded and uses TCP sockets with the TLS extensions; the server server socket is managed by one thread and the client socket is managed by the main thread.

My reference(s) for the application were the `echo_server` / `echo_client` sample applications and I'm just using the certificates / keys from those sample applications as-is. However, I'm experiencing a handshake error in TLS.

**Describe the solution you'd like**
It might be useful to have it a test that verified this kind of functionality in Zephyr CI

**Describe alternatives you've considered**
Have looked around for existing tests that exercise this particular setup, but have found none.

**Additional context**

### PR DESCRIPTION
**Pull request by cfriedt**
This change adds a test for the socket API TLS extensions.

Fixes #30936

Sorry for the duplicate of #30942 but I had some repository issues and had to recreate the PR.

### CODE REVIEW

**Review by rlubos — APPROVED** [review]
I see that my comments from the original PR were addressed, LGTM.

File: tests/net/socket/tls/src/main.c [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+	__ASSERT(r != -1, "accept() failed (%d)", r);
+	client_fd = r;
+
+	memset(addrstr, '\0', sizeof(addrstr));
+	addrstrp = (char *)inet_ntop(AF_INET, &sa.sin_addr,
+	addrstr, sizeof(addrstr));
+	__ASSERT(addrstrp != NULL, "inet_ntop() failed (%d)", errno);
+
+	NET_DBG("accepted connection from [%s]:%d as fd %d",
+	log_strdup(addrstr), ntohs(sa.sin_port), client_fd);
```
Comment (jukkar) [inline tests/net/socket/tls/src/main.c:175]: Indentation wrong

File: tests/net/socket/tls/src/main.c [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+
+	NET_DBG("Server thread running");
+
+	memset(&sa, 0, sizeof(sa));
+	addrlen = sizeof(sa);
+
+	NET_DBG("Accepting client connection..");
+	k_sem_give(&server_sem);
+	r = accept(server_fd, (struct sockaddr *)&sa, &addrlen);
+	__ASSERT(r != -1, "accept() failed (%d)", r);
```
Comment (jukkar) [inline tests/net/socket/tls/src/main.c:166]: You should use `zassert*()` macros like in other functions in this file.

File: tests/net/socket/tls/src/main.c [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+			break;
+		default:
+			zassert_true(false,
+				"unrecognized TLS peer verify type %d",
+				peer_verify);
+			return;
+		}
+
+		r = setsockopt(server_fd, SOL_TLS, TLS_SEC_TAG_LIST,
+			sec_tag_list, sec_tag_list_size);
```
Comment (jukkar) [inline tests/net/socket/tls/src/main.c:261]: indentation

File: tests/net/socket/tls/src/main.c [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+		zassert_not_equal(r, -1, "failed to set TLS_SEC_TAG_LIST (%d)",
+			errno);
```
Comment (jukkar) [inline tests/net/socket/tls/src/main.c:263]: ditto

File: tests/net/socket/tls/src/main.c [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+
+		r = setsockopt(server_fd, SOL_TLS, TLS_HOSTNAME, "localhost",
+			sizeof("localhost"));
```
Comment (jukkar) [inline tests/net/socket/tls/src/main.c:266]: and here

File: tests/net/socket/tls/src/main.c [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+		zassert_not_equal(r, -1, "failed to set TLS_HOSTNAME (%d)",
+			errno);
```
Comment (jukkar) [inline tests/net/socket/tls/src/main.c:268]: and here. Actually I see indentation issues all over the place. Could you go through this file and fix them all in one go, I do not bother comment them all.

**Review by jukkar — COMMENTED** [review]
Tests themselves look good, just minor nit about style issues.

**Comment by ceolin** [comment]
```
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.x2apic [bit 21]
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..*** Booting Zephyr OS build zephyr-v2.4.0-3359-g971bb4a91543  ***
W: No entropy device on the system, TLS communication may be insecure!
I: Initializing network
I: IPv4 address: 192.0.2.1
Running test suite tls_socket_api_extension
===================================================================
START - test_tls_peer_verify_none
 PASS - test_tls_peer_verify_none
===================================================================
START - test_tls_peer_verify_optional
E: TLS handshake error: -4290
E: TLS handshake error: -7280

    Assertion failed at ../src/main.c:366: test_common: (r equal to -1)
failed to connect (53)
 FAIL - test_tls_peer_verify_optional
===================================================================
START - test_tls_peer_verify_required
E: Identical connection handler 0x139150 already found.

    Assertion failed at ../src/main.c:281: test_common: (r equal to -1)
failed to listen (69)
 FAIL - test_tls_peer_verify_required
===================================================================
Test suite tls_socket_api_extension failed.
===================================================================
PROJECT EXECUTION FAILED
ASSERTION FAIL [r != -1] @ ../src/main.c:166
	accept() failed (-1)
E: RAX: 0x0000000000000004 RBX: 0x0000000000000000 RCX: 0x00000000000003f8 RDX: 0x0000000000130930
E: RSI: 0x00000000000000a6 RDI: 0x00000000001308b8 RBP: 0x0000000000147fe0 RSP: 0x0000000000147fa8
E:  R8: 0x0000000000000001  R9: 0x0000000000147e2d R10: 0x00000000ffffffff R11: 0x0000000000000031
E: R12: 0x00000000ffffffff R13: 0x0000000000000246 R14: 0x0000000000000000 R15: 0x0000000000000000
E: RSP: 0x0000000000147fa8 RFLAGS: 0x0000000000000206 CS: 0x0018 CR3: 0x000000000018b000
E: RIP: 0x00000000001023fa
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x135000 (server)
E: Halting system

File: tests/net/socket/tls/src/main.c (hunk 1 above)
Comment (cfriedt) [reply tests/net/socket/tls/src/main.c:175]: I'm not seeing that on my end.
```
tests/net/socket/tls/src/main.c total: 0 errors, 0 warnings, 475 lines checked

tests/net/socket/tls/src/main.c has no obvious style problems and is ready for submission.
```

File: tests/net/socket/tls/src/main.c (hunk 2 above)
Comment (cfriedt) [reply tests/net/socket/tls/src/main.c:166]: Does that work properly when called from a separate thread? I vaguely remember that the test did not fail reliably when I used `zassert()` here.

**Comment by cfriedt** [comment]
> ```
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.x2apic [bit 21]
> SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
> Booting from ROM..*** Booting Zephyr OS build zephyr-v2.4.0-3359-g971bb4a91543  ***
> W: No entropy device on the system, TLS communication may be insecure!
> I: Initializing network
> I: IPv4 address: 192.0.2.1
> Running test suite tls_socket_api_extension
> ===================================================================
> START - test_tls_peer_verify_none
>  PASS - test_tls_peer_verify_none
> ===================================================================
> START - test_tls_peer_verify_optional
> E: TLS handshake error: -4290
> E: TLS handshake error: -7280
> 
>     Assertion failed at ../src/main.c:366: test_common: (r equal to -1)
> failed to connect (53)
>  FAIL - test_tls_peer_verify_optional
> ===================================================================
> START - test_tls_peer_verify_required
> E: Identical connection handler 0x139150 already found.
> 
>     Assertion failed at ../src/main.c:281: test_common: (r equal to -1)
> failed to listen (69)
>  FAIL - test_tls_peer_verify_required
> ===================================================================
> Test suite tls_socket_api_extension failed.
> ===================================================================
> PROJECT EXECUTION FAILED
> ASSERTION FAIL [r != -1] @ ../src/main.c:166
> 	accept() failed (-1)
> E: RAX: 0x0000000000000004 RBX: 0x0000000000000000 RCX: 0x00000000000003f8 RDX: 0x0000000000130930
> E: RSI: 0x00000000000000a6 RDI: 0x00000000001308b8 RBP: 0x0000000000147fe0 RSP: 0x0000000000147fa8
> E:  R8: 0x0000000000000001  R9: 0x0000000000147e2d R10: 0x00000000ffffffff R11: 0x0000000000000031
> E: R12: 0x00000000ffffffff R13: 0x0000000000000246 R14: 0x0000000000000000 R15: 0x0000000000000000
> E: RSP: 0x0000000000147fa8 RFLAGS: 0x0000000000000206 CS: 0x0018 CR3: 0x000000000018b000
> E: RIP: 0x00000000001023fa
> E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
> E: Current thread: 0x135000 (server)
> E: Halting system
> ```

I've run into that too. It seemed to happen after I made changes requested in #30942. Not 100% sure why this is happening, so it will take some investigation.

**Comment by rlubos** [comment]
> I've run into that too. It seemed to happen after I made changes requested in #30942. Not 100% sure why this is happening, so it will take some investigation.

It seems that `qemu_x86_64` needs more heap memory. Not trivial to figure out from the mbedTLS logs, as there's no error code indicated by `E: TLS handshake error: -4290`, but it's most likely a sum of `MBEDTLS_ERR_MPI_ALLOC_FAILED` and `MBEDTLS_ERR_RSA_PUBLIC_FAILED` (there are places where they sum error codes :/). Anyway, adding 2kB of mbedTLS heap solved the issue for me.

**Comment by cfriedt** [comment]
Thanks @rlubos - that solved the issue ;-)

File: tests/net/socket/tls/src/main.c [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+
+/**
+ * @brief The server thread function
+ *
+ * This function simply accepts a client connection and
+ * echoes the first @ref SECRET_SIZE bytes of the first
+ * packet. After that, the server is closed and connections
+ * are no longer accepted.
+ *
+ * @param arg0 a pointer to the int representing the server filed descriptor
```
Comment (cfriedt) [inline tests/net/socket/tls/src/main.c:141]: Fix typo

File: tests/net/socket/tls/src/main.c (hunk 2 above)
Comment (cfriedt) [reply tests/net/socket/tls/src/main.c]: Works fine, my mistake.

File: tests/net/socket/tls/testcase.yaml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+common:
+  tags: net socket tls
+  # these platforms exhibited out of memory conditions
```
Comment (jukkar) [inline tests/net/socket/tls/testcase.yaml:3]: If we have min_ram setting below, is it really necessary to blacklist certain platforms?

Comment (cfriedt) [reply tests/net/socket/tls/testcase.yaml:3]: I'll give it a shot without shortly.

**Comment by cfriedt** [comment]
I moved my tests from `/tls` to `/tls_ext` because there seemed to be way too many conflicts to resolve in a timely manner (`main.c`, `prj.conf`, `testcase.yml` - basically everything)

**Review by jukkar — APPROVED** [review]

**Pull request merged**